src/dune_lang/string_with_vars.mli

```diff
@@ -114,6 +114,20 @@ module type Expander = sig
     -> f:(Value.t list, 'error) result app expander
     -> ('a, 'error) result app
 
+  (** Similar to [expand_result]. Returns a [`Value _] if [t] is a single
+      unquoted pform or string containing no pforms. In all other cases (quoted
+      pforms and strings made up of a mix of pforms and text) then a
+      [`Deferred_concat _] is returned containing each part of the result. This
+      preserves the [Value.t] of each pform rather than collapsing path values
+      to strings (as done by [expand_result] and [expand]). This is why
+      [expand_result_deferred_concat] does not need a [dir] argument. *)
+  val expand_result_deferred_concat
+    :  t
+    -> mode:'a Mode.t
+    -> f:(Value.t list, 'error) result app expander
+    -> ([ `Values of 'a | `Deferred_concat of Value.Deferred_concat.t ], 'error) result
```

Comment:
The type signature doesn't look right to me. The index in `'a Mode.t` shouldn't be discarded in `Value.Deferred_concat.t`. We should be able to only expand `Value.Deferred_concat.t` to a list of values or a single value based on the `'a Mode.t`.

Reply:
> The index in 'a Mode.t shouldn't be discarded in Value.Deferred_concat.t
It's safe to discard the `'a` because regardless of `'a`, `Deferred_concat` will only be returned in cases where only a single value would have been returned anyway.

> We should be able to only expand Value.Deferred_concat.t to a list of values or a single value based on the 'a Mode.t.
That doesn't sound right to me. A `Deferred_concat.t` always represents a single value. If `'a` is `Multi` then `%{targets}` would expand to all the targets as separate values (say `foo.txt` and `bar.txt`) whereas `"%{targets}` would expand to the single value `foo.txt bar.txt`. All cases involving concatenation expand to a single value and the index `'a` shouldn't affect the treatment of such values.

Reply:
Okay, indeed `Deferred_concat.t` is possible for all modes. It's rather confusing though.

Reply:
Yeh this function's type is a bit messy. I think making values deferred by default will simplify it.